meta2v2/meta2_utils.c

```diff
@@ -2245,15 +2322,83 @@ GError *m2db_check_content(GSList *beans, struct namespace_info_s *nsinfo,
 			err = NEWERROR(CODE_CONTENT_CORRUPTED, "Content broken and irreparable");
 			break;
 	}
+	return err;
+}
+
+void
+m2db_check_content_quality(struct m2v2_sorted_content_s *sorted_content,
+		GSList *chunk_meta, GSList **to_be_improved)
+{
+	EXTRA_ASSERT(to_be_improved != NULL);
+	GHashTable *chunk_items = g_hash_table_new_full(g_str_hash, g_str_equal,
+			g_free, g_free);
+	_load_chunk_qualities(chunk_items, chunk_meta);
+
+	gboolean _on_metachunk(gpointer ppos, GSList *chunks, gpointer udata UNUSED) {
+		gint mc_pos = GPOINTER_TO_INT(ppos);
+		GString *out = g_string_sized_new(1024);
+		g_string_append_c(out, '{');
+		oio_str_gstring_append_json_pair_int(out, "pos", mc_pos);
+		g_string_append_c(out, ',');
+		oio_str_gstring_append_json_quote(out, "chunks");
+		g_string_append(out, ":[");
+		gboolean must_send_event = FALSE;
+		for (GSList *cur = chunks; cur; cur = cur->next) {
+			GString *chunk_id = CHUNKS_get_id(cur->data);
+			if (cur != chunks)
+				g_string_append_c(out, ',');
+			struct oio_lb_selected_item_s *item = g_hash_table_lookup(
+					chunk_items, chunk_id->str);
+			if (item) {
+				g_string_append_c(out, '{');
+				meta2_json_encode_bean(out, cur->data);
+				g_string_append_c(out, ',');
+				oio_str_gstring_append_json_quote(out, "quality");
+				g_string_append_c(out, ':');
+				_selected_item_quality_to_json(out, item);
+				if (item->final_dist <= item->warn_dist ||
+						strcmp(item->final_slot, item->expected_slot)) {
+					must_send_event = TRUE;
+				}
+				g_string_append_c(out, '}');
+			}
```

Comment:
if `item` was null on following loops, JSON will be invalid ?
IMO l.2349 should be after  l.2353 ?